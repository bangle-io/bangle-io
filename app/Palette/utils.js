import React from 'react';

export function addBoldToTitle(string, query) {
  if (!query) {
    return string;
  }

  let newString = string.split(query);

  return newString.flatMap((r, i) => {
    if (i === 0) {
      return r;
    }
    return [<b key={i}>{query}</b>, r];
  });
}
