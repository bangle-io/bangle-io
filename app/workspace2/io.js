import * as idb from 'idb-keyval';

export class FileOps {
  // file
  static async getFile(wsPath) {
    return idb.get(wsPath);
  }

  static async deleteFile(wsPath) {
    const file = await this.getFile(wsPath);
    if (!file) {
      throw new FileError(`File ${wsPath} does not exists`);
    }
    return idb.del(wsPath);
  }

  static async renameFile(wsPath, newWsPath) {
    const file = await this.getFile(wsPath);

    if (!file) {
      throw new FileError(`File ${wsPath} does not exists`);
    }

    const newFile = await this.getFile(newWsPath);

    if (newFile) {
      throw new FileError(`File ${newWsPath} already exists`);
    }

    await this.deleteFile(wsPath);
    return this.createFile(newWsPath, file);
  }

  static async createFile(wsPath, payload) {
    if (await this.getFile(wsPath)) {
      throw new FileError(`File ${wsPath} already exists`);
    }
    return idb.set(wsPath, payload);
  }

  static async updateFile(wsPath, payload) {
    if (!(await this.getFile(wsPath))) {
      throw new FileError(`File ${wsPath} does not exists`);
    }
    return idb.set(wsPath, payload);
  }

  static async listFiles(wsName) {
    const keys = (await idb.keys()) || [];

    return keys.filter((k) => k.startsWith(wsName + ':'));
  }
}

export class FileError extends Error {
  constructor(message) {
    super('FileError: ' + message);
    this.name = 'FileError';
  }
}
export class FilePermissionError extends Error {
  constructor(message) {
    super('FilePermissionError: ' + message);
    this.name = 'FilePermissionError';
  }
}
