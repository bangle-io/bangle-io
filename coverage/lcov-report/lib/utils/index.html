
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for lib/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> lib/utils</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">31.72% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>46/145</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">21.53% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>14/65</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">14.7% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>5/34</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">31.25% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>45/144</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="cx.ts"><a href="cx.ts.html">cx.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file low" data-value="safe-js.ts"><a href="safe-js.ts.html">safe-js.ts</a></td>
	<td data-value="25.88" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 25%"></div><div class="cover-empty" style="width: 75%"></div></div>
	</td>
	<td data-value="25.88" class="pct low">25.88%</td>
	<td data-value="85" class="abs low">22/85</td>
	<td data-value="28.26" class="pct low">28.26%</td>
	<td data-value="46" class="abs low">13/46</td>
	<td data-value="7.69" class="pct low">7.69%</td>
	<td data-value="26" class="abs low">2/26</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="84" class="abs low">21/84</td>
	</tr>

<tr>
	<td class="file low" data-value="window.ts"><a href="window.ts.html">window.ts</a></td>
	<td data-value="28" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 28%"></div><div class="cover-empty" style="width: 72%"></div></div>
	</td>
	<td data-value="28" class="pct low">28%</td>
	<td data-value="50" class="abs low">14/50</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="18" class="abs low">0/18</td>
	<td data-value="28.57" class="pct low">28.57%</td>
	<td data-value="7" class="abs low">2/7</td>
	<td data-value="28" class="pct low">28%</td>
	<td data-value="50" class="abs low">14/50</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-12-02T21:35:16.931Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    