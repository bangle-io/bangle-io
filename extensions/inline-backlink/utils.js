export function removeMdExtension(str) {
  if (str.endsWith('.md')) {
    return str.slice(0, -3);
  }
  return str;
}
