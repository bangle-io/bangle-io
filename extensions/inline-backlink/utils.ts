import {
  filePathToWsPath,
  removeExtension,
  resolvePath,
  sanitizeFilePath,
  suffixWithNoteExtension,
  validateWsPath,
} from '@bangle.io/ws-path';

export function wsPathFromQuery(query: string, wsName: string) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = suffixWithNoteExtension(filePath);
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}

export function getBacklinkPath(wsPath: string, allWsPaths: readonly string[]) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeExtension(filePath);
  }

  return removeExtension(fileName);
}
