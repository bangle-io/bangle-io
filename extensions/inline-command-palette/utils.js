export function replaceSuggestText(newText) {
  return (state, dispatch, view) => {};
}
