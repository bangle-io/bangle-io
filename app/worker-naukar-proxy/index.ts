import { mainInjectAbortableProxy } from '@bangle.io/abortable-worker';
import type { NaukarWorkerAPI } from '@bangle.io/shared-types';

import { waitForProxy } from './wait-for-proxy';

const naukarRef: { current: undefined | NaukarWorkerAPI } = {
  current: undefined,
};

const check: Array<() => void> = [];

let isNaukarReady = {
  current: false,
};

const editorProxy = waitForProxy<NaukarWorkerAPI['editor']>(
  isNaukarReady,
  () => {
    return naukarRef.current?.editor!;
  },
);

check.push(editorProxy.check);

const testProxy = waitForProxy<NaukarWorkerAPI['test']>(isNaukarReady, () => {
  return naukarRef.current?.test!;
});

check.push(testProxy.check);

const workspaceProxy = waitForProxy<NaukarWorkerAPI['workspace']>(
  isNaukarReady,
  () => {
    return naukarRef.current?.workspace!;
  },
);

check.push(workspaceProxy.check);

const abortableProxy = waitForProxy<NaukarWorkerAPI['abortable']>(
  isNaukarReady,
  () => {
    return naukarRef.current?.abortable!;
  },
);

check.push(abortableProxy.check);

const replicaSlicesProxy = waitForProxy<NaukarWorkerAPI['replicaSlices']>(
  isNaukarReady,
  () => {
    return naukarRef.current?.replicaSlices!;
  },
);

check.push(replicaSlicesProxy.check);

if (check.length !== 5) {
  throw new Error('Incorrect number of checks');
}

export function _setWorker(incomingNaukar: NaukarWorkerAPI) {
  naukarRef.current = incomingNaukar;
  isNaukarReady.current = true;

  check.forEach((check) => check());
}

export function _clearWorker() {
  naukarRef.current = undefined;
  isNaukarReady.current = false;
}

// a proxy will stall any methods
// until the worker is marked ready.
export const naukarProxy: NaukarWorkerAPI = {
  editor: editorProxy.proxy,

  test: testProxy.proxy,

  workspace: workspaceProxy.proxy,

  abortable: mainInjectAbortableProxy(abortableProxy.proxy),

  replicaSlices: replicaSlicesProxy.proxy,
};
