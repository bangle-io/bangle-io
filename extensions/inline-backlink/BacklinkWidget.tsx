import React, { useCallback, useEffect, useMemo, useState } from 'react';

import { nsmApi2 } from '@bangle.io/api';
import { isAbortError } from '@bangle.io/mini-js-utils';
import type { HighlightTextType, SearchMatch } from '@bangle.io/search-pm-node';
import {
  ButtonIcon,
  ChevronDownIcon,
  ChevronRightIcon,
  Sidebar,
} from '@bangle.io/ui-components';
import {
  cx,
  getMouseClickType,
  MouseClick,
  safeCancelIdleCallback,
  safeRequestIdleCallback,
} from '@bangle.io/utils';
import { naukarProxy } from '@bangle.io/worker-naukar-proxy';
import { createWsPath, removeExtension, resolvePath } from '@bangle.io/ws-path';

const IconStyle = {
  height: 16,
  width: 16,
};

interface BacklinkSearchResult {
  wsPath: string;
  matches: SearchMatch[];
}

export function BacklinkWidget() {
  const backlinkSearchResult = useBacklinkSearch();

  const makeOnClick = useCallback((_path?: string) => {
    return (event: React.MouseEvent<any>) => {
      if (!_path) {
        return;
      }

      const wsPath = createWsPath(_path);

      const clickType = getMouseClickType(event);

      if (clickType === MouseClick.NewTab) {
        nsmApi2.workspace.openWsPathInNewTab(wsPath);
      } else if (clickType === MouseClick.ShiftClick) {
        nsmApi2.workspace.pushSecondaryWsPath(wsPath);
      } else {
        nsmApi2.workspace.pushPrimaryWsPath(wsPath);
      }
    };
  }, []);

  const [openedItems, updateOpenedItems] = useState(() => new Set<string>());
  const isCollapsed = useCallback(
    (r: BacklinkSearchResult) => {
      return !openedItems.has(r.wsPath);
    },
    [openedItems],
  );

  return (
    <div
      data-testid="inline-backlink_widget-container"
      className="flex flex-col"
    >
      {!backlinkSearchResult || backlinkSearchResult.length === 0 ? (
        <span>
          üêí No backlinks found!
          <br />
          <span className="font-light">
            Create one by typing <kbd className="font-normal">[[</kbd> followed
            by the name of the note.
          </span>
        </span>
      ) : (
        backlinkSearchResult.map((r, i) => {
          return (
            <React.Fragment key={i}>
              <Sidebar.Row2
                titleClassName=""
                className={'rounded text-sm truncate py-1 select-none'}
                extraInfoClassName="ml-1 text-sm"
                onClick={makeOnClick(r.wsPath)}
                item={{
                  uid: r.wsPath,
                  showDividerAbove: false,
                  title: resolvePath(r.wsPath).fileNameWithoutExt,
                  leftNode: (
                    <ButtonIcon
                      onClick={(e) => {
                        updateOpenedItems((items) => {
                          if (items.has(r.wsPath)) {
                            const clone = new Set(items);
                            clone.delete(r.wsPath);

                            return clone;
                          } else {
                            const clone = new Set(items);
                            clone.add(r.wsPath);

                            return clone;
                          }
                        });
                        e.stopPropagation();
                        e.preventDefault();
                      }}
                    >
                      {isCollapsed(r) ? (
                        <ChevronRightIcon style={IconStyle} />
                      ) : (
                        <ChevronDownIcon style={IconStyle} />
                      )}
                    </ButtonIcon>
                  ),
                  rightNode: (
                    <ButtonIcon className="text-xs font-semibold rounded bg-colorNeutralSolidFaint px-1">
                      {r.matches.length}
                    </ButtonIcon>
                  ),
                  rightHoverNode: (
                    <ButtonIcon className="text-xs font-semibold rounded bg-colorNeutralSolidFaint px-1">
                      {r.matches.length}
                    </ButtonIcon>
                  ),
                }}
              />
              {!isCollapsed(r) &&
                r.matches.map((matchObj, j) => (
                  <Sidebar.Row2
                    key={j}
                    className={
                      'B-search-notes_search-result-text-match ml-1 pl-3 rounded ' +
                      (j === 0 ? 'mt-0' : 'mt-1')
                    }
                    onClick={makeOnClick(r.wsPath)}
                    titleClassName="text-sm"
                    item={{
                      uid: 'search-result-text-match-' + j,
                      title: <HighlightText highlightText={matchObj.match} />,
                    }}
                  ></Sidebar.Row2>
                ))}
            </React.Fragment>
          );
        })
      )}
    </div>
  );
}

function useBacklinkSearch(): BacklinkSearchResult[] | undefined {
  const { focusedEditorId } = nsmApi2.editor.useEditor();
  const { wsName, openedWsPaths } = nsmApi2.workspace.useWorkspace();

  const [results, updateResults] = useState<BacklinkSearchResult[] | undefined>(
    undefined,
  );

  const calculateResult = useCallback(
    (focusedWsPath: string, controller: AbortController) => {
      if (!wsName) {
        return;
      }
      naukarProxy.abortable
        .abortableSearchWsForPmNode(controller.signal, wsName, 'backlink:*', [
          {
            nodeName: 'wikiLink',
            dataAttrName: 'path',
            printBefore: '[[',
            printAfter: ']]',
            queryIdentifier: 'backlink:',
          },
        ])
        .then(
          (result) => {
            const fileName = removeExtension(
              resolvePath(focusedWsPath).fileName,
            );

            updateResults(
              result
                .map((r) => {
                  const newMatches = r.matches.filter((match) => {
                    const [, highlightTextMatch] = match.match;

                    if (highlightTextMatch) {
                      return highlightTextMatch.includes(fileName);
                    }

                    return false;
                  });

                  return {
                    wsPath: r.uid,
                    matches: newMatches,
                  };
                })
                .filter((r) => r.matches.length > 0),
            );
          },
          (error) => {
            if (isAbortError(error)) {
              return;
            }

            throw error;
          },
        );
    },
    [wsName],
  );

  const focusedWsPath = useMemo(() => {
    if (focusedEditorId != null) {
      return openedWsPaths.getByIndex(focusedEditorId);
    }

    return undefined;
  }, [openedWsPaths, focusedEditorId]);

  useEffect(() => {
    const controller = new AbortController();
    let cb =
      focusedWsPath &&
      safeRequestIdleCallback(
        () => {
          calculateResult(focusedWsPath, controller);
        },
        { timeout: 1000 },
      );

    return () => {
      controller.abort();

      if (cb) {
        safeCancelIdleCallback(cb);
      }
    };
  }, [focusedWsPath, wsName, calculateResult]);

  return results;
}

function HighlightText({
  highlightText,
}: {
  highlightText: HighlightTextType;
}) {
  return (
    <div className="B-inline-backlink_highlight-text-container">
      {highlightText.map((t, i) => (
        <span
          key={i}
          className={cx(i % 2 === 1 && 'bg-colorPromoteSolidFaint', 'text-sm')}
        >
          {t}
        </span>
      ))}
    </div>
  );
}
