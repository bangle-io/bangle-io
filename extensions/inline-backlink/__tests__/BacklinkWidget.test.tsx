/**
 * @jest-environment jsdom
 */
import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import { useEditorManagerContext } from '@bangle.io/slice-editor-manager';
import { pushWsPath, useWorkspaceContext } from '@bangle.io/slice-workspace';
import {
  getUseEditorManagerContextReturn,
  getUseWorkspaceContextReturn,
} from '@bangle.io/test-utils/function-mock-return';
import { sleep } from '@bangle.io/utils';
import { naukarProxy } from '@bangle.io/worker-naukar-proxy';
import { OpenedWsPaths } from '@bangle.io/ws-path';

import { BacklinkWidget } from '../BacklinkWidget';

jest.mock('@bangle.io/slice-workspace');

jest.mock('@bangle.io/slice-editor-manager');

jest.mock('@bangle.io/worker-naukar-proxy', () => {
  return {
    naukarProxy: {
      abortableSearchWsForPmNode: jest.fn(),
    },
  };
});

const pushWsPathMock = pushWsPath as jest.MockedFunction<typeof pushWsPath>;

let useWorkspaceContextMock = useWorkspaceContext as jest.MockedFunction<
  typeof useWorkspaceContext
>;

let useEditorManagerContextMock =
  useEditorManagerContext as jest.MockedFunction<
    typeof useEditorManagerContext
  >;

let abortableSearchWsForPmNodeMock =
  naukarProxy.abortableSearchWsForPmNode as jest.MockedFunction<
    typeof naukarProxy.abortableSearchWsForPmNode
  >;

beforeEach(() => {
  abortableSearchWsForPmNodeMock.mockImplementation(async () => []);

  useWorkspaceContextMock.mockImplementation(() => {
    return {
      ...getUseWorkspaceContextReturn,
      wsName: 'test-back-ws',
    };
  });

  pushWsPathMock.mockImplementation(() => () => true);

  useEditorManagerContextMock.mockImplementation(() => {
    return {
      ...getUseEditorManagerContextReturn,
    };
  });
});

test('renders with blank data', async () => {
  const renderResult = render(
    <div>
      <BacklinkWidget />
    </div>,
  );

  expect(renderResult.container.innerHTML).toContain(`No backlinks found`);
  expect(renderResult.container).toMatchInlineSnapshot(`
    <div>
      <div>
        <div
          class="inline-backlink_widget-container flex flex-col"
        >
          <span>
            üêí No backlinks found!
            <br />
            <span
              class="font-light"
            >
              Create one by typing 
              <kbd
                class="font-normal"
              >
                [[
              </kbd>
               followed by the name of the note.
            </span>
          </span>
        </div>
      </div>
    </div>
  `);
});

test('handles abort error', async () => {
  abortableSearchWsForPmNodeMock.mockImplementation(async () => {
    throw new DOMException('Aborted', 'AbortError');
  });
  const openedWsPaths = new OpenedWsPaths([
    'test-back-ws:my-linked-note-1.md',
    undefined,
  ]);

  const pushWsPath = jest.fn();

  useWorkspaceContextMock.mockImplementation(() => {
    return {
      ...getUseWorkspaceContextReturn,
      wsName: 'test-back-ws',
      openedWsPaths,
      pushWsPath,
    };
  });

  useEditorManagerContextMock.mockImplementation(() => {
    return {
      ...getUseEditorManagerContextReturn,
      focusedEditorId: 0,
    };
  });

  const renderResult = render(
    <div>
      <BacklinkWidget />
    </div>,
  );

  await sleep(15);

  expect(abortableSearchWsForPmNodeMock).toBeCalledTimes(1);
  expect(renderResult.container.innerHTML).toContain(`No backlinks found`);
  expect(renderResult.container).toMatchInlineSnapshot(`
    <div>
      <div>
        <div
          class="inline-backlink_widget-container flex flex-col"
        >
          <span>
            üêí No backlinks found!
            <br />
            <span
              class="font-light"
            >
              Create one by typing 
              <kbd
                class="font-normal"
              >
                [[
              </kbd>
               followed by the name of the note.
            </span>
          </span>
        </div>
      </div>
    </div>
  `);
});

test('renders backlinks', async () => {
  // We have opened note-1
  // which is referenced by note-2

  const openedWsPaths = new OpenedWsPaths([
    'test-back-ws:my-linked-note-1.md',
    undefined,
  ]);

  useWorkspaceContextMock.mockImplementation(() => {
    return {
      ...getUseWorkspaceContextReturn,
      wsName: 'test-back-ws',
      openedWsPaths,
    };
  });

  useEditorManagerContextMock.mockImplementation(() => {
    return {
      ...getUseEditorManagerContextReturn,
      focusedEditorId: 0,
    };
  });

  abortableSearchWsForPmNodeMock.mockImplementation(async () => {
    return [
      {
        matches: [
          {
            match: ['', '[[my-linked-note-1]]', ' some content'],
            parent: 'doc',
            parentPos: 2,
          },
        ],
        // reference to 'note-1' in 'note-2' note
        uid: 'test-back-ws:my-linked-note-2.md',
      },
    ];
  });

  const renderResult = render(
    <div>
      <BacklinkWidget />
    </div>,
  );

  let targetOption;
  await waitFor(() => {
    targetOption = renderResult.container.querySelector(
      '[data-id="test-back-ws:my-linked-note-2.md"]',
    );
    expect(targetOption).toBeTruthy();
  });

  expect(renderResult.container.innerHTML).toContain(`my-linked-note-2`);
  expect(renderResult.container).toMatchSnapshot();

  const backlinkedPage = renderResult.container.querySelector(
    '[data-id="test-back-ws:my-linked-note-2.md"]',
  );
  expect(abortableSearchWsForPmNodeMock).toBeCalledTimes(1);

  await fireEvent.click(backlinkedPage!);

  expect(pushWsPath).toBeCalledTimes(1);
  expect(pushWsPath).nthCalledWith(
    1,
    'test-back-ws:my-linked-note-2.md',
    false,
    false,
  );

  const expandButton = renderResult.container.querySelector(
    '[data-id="test-back-ws:my-linked-note-2.md"] button',
  );

  await fireEvent.click(expandButton!);

  expect(pushWsPath).toBeCalledTimes(1);

  expect(renderResult.container.innerHTML).toContain('[[my-linked-note-1]]');

  expect(
    renderResult.container.querySelector('.highlight-text-container'),
  ).toMatchSnapshot();

  await fireEvent.click(
    renderResult.container.querySelector('.highlight-text-container')!,
  );
  expect(pushWsPath).toBeCalledTimes(2);
  expect(pushWsPath).nthCalledWith(
    2,
    'test-back-ws:my-linked-note-2.md',
    false,
    false,
  );
});
