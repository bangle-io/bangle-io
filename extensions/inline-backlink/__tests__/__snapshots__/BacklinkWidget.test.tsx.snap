// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders backlinks 1`] = `
<div>
  <div>
    <div
      class="flex flex-col"
      data-testid="inline-backlink_widget-container"
    >
      <div
        class="B-ui-components_sidebar-row2 BU_hover rounded text-sm truncate py-1 select-none "
        data-id="test-back-ws:my-linked-note-2.md"
        role="button"
        style="cursor: pointer; display: flex; justify-content: space-between; user-select: none;"
      >
        <div
          class="flex flex-row"
        >
          <div
            class="B-ui-components_left-node"
          >
            <button
              class="focus:outline-none "
              data-bangle-editor-break="true"
              data-bangle-editor-pos="bottom"
              type="button"
            >
              <svg
                fill="currentColor"
                style="height: 16px; width: 16px;"
                viewBox="0 0 18 18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#ff13dc"
                  height="18"
                  opacity="0"
                  width="18"
                />
                <path
                  d="M12,9a.994.994,0,0,1-.2925.7045l-3.9915,3.99a1,1,0,1,1-1.4355-1.386l.0245-.0245L9.5905,9,6.3045,5.715A1,1,0,0,1,7.691,4.28l.0245.0245,3.9915,3.99A.994.994,0,0,1,12,9Z"
                />
              </svg>
            </button>
          </div>
          <span
            class=""
          >
            <span
              class=""
            >
              my-linked-note-2
            </span>
          </span>
        </div>
        <div
          class="flex flex-row"
        >
          <span
            class="B-ui-components_right-node"
          >
            <button
              class="focus:outline-none text-xs font-semibold rounded bg-colorNeutralSolidFaint px-1 "
              data-bangle-editor-break="true"
              data-bangle-editor-pos="bottom"
              type="button"
            >
              1
            </button>
          </span>
          <span
            class="B-ui-components_right-hover-node"
          >
            <button
              class="focus:outline-none text-xs font-semibold rounded bg-colorNeutralSolidFaint px-1 "
              data-bangle-editor-break="true"
              data-bangle-editor-pos="bottom"
              type="button"
            >
              1
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders backlinks 2`] = `
<div
  class="B-inline-backlink_highlight-text-container"
>
  <span
    class="text-sm "
  />
  <span
    class="bg-colorPromoteSolidFaint text-sm "
  >
    [[my-linked-note-1]]
  </span>
  <span
    class="text-sm "
  >
     some content
  </span>
</div>
`;
